File: src/apps/gin_scoring/domain/commands/_save_game_result.py
import typing as t

from django.utils.timezone import now

from ...domain.gin_rummy import GAME_OUTCOME, calculate_round_score
from ...models import GameResult


def save_game_result(
    player_north_name: str,
    player_south_name: str,
    outcome: GAME_OUTCOME,
    winner_name: t.Optional[str],
    deadwood_value: int,
) -> GameResult:
    is_draw = outcome == "draw"

    winner_score = None
    if is_draw:
        winner_name = None
    else:
        winner_name = winner_name
        winner_score = calculate_round_score(game_outcome=outcome, deadwood_value=deadwood_value)

    game_result_model = GameResult(
        player_north_name=player_north_name,
        player_south_name=player_south_name,
        outcome=outcome,
        winner_name=winner_name,
        deadwood_value=deadwood_value,
        winner_score=winner_score,
        created_at=now(),
    )
    game_result_model.save()

    return game_result_model

File: internal/domain/mutations/save_game_result.go
package mutations

import (
	"context"
	"fmt"
	"github.com/drbenton/gin-scoring/internal"
	"github.com/drbenton/gin-scoring/internal/domain"
	"github.com/drbenton/gin-scoring/internal/models"
	"github.com/volatiletech/null/v8"
	"github.com/volatiletech/sqlboiler/v4/boil"
)

type SaveGameResultCommand struct {
	PlayerNorthName string
	PlayerSouthName string
	Outcome         domain.GameOutcome
	WinnerName      *string
	DeadwoodValue   uint
}

func SaveGameResult(ctx context.Context, cmd SaveGameResultCommand) (*models.GameResult, error) {
	err := checkSaveGameResultInput(cmd)
	if err != nil {
		return nil, err
	}

	winnerScore, err := domain.CalculateRoundScore(cmd.Outcome, cmd.DeadwoodValue)
	if err != nil {
		return nil, err
	}

	resultModel := models.GameResult{
		PlayerNorthName: cmd.PlayerNorthName,
		PlayerSouthName: cmd.PlayerSouthName,
		Outcome:         string(cmd.Outcome),
		DeadwoodValue:   int16(cmd.DeadwoodValue),
		WinnerName:      null.StringFromPtr(cmd.WinnerName),
		WinnerScore:     null.Int16From(int16(winnerScore)),
	}
	err = resultModel.Insert(ctx, internal.DB(), boil.Infer())
	if err != nil {
		return nil, err
	}
	return &resultModel, nil
}

func checkSaveGameResultInput(cmd SaveGameResultCommand) error {
	winnerName := *cmd.WinnerName
	if cmd.Outcome != domain.GameOutcomeDraw && winnerName != cmd.PlayerNorthName && winnerName != cmd.PlayerSouthName {
		return fmt.Errorf("winner name '%s' is neither '%s' or '%s'", winnerName, cmd.PlayerNorthName, cmd.PlayerSouthName)
	}
	return nil
}
